bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/ModbusPollerThingHandler.java

```diff
@@ -61,7 +63,8 @@
      * @author Sami Salonen
      *
      */
-    private class ReadCallbackDelegator implements ModbusReadCallback {
+    private class ReadCallbackDelegator
+            implements ModbusReadCallback, ModbusFailureCallback<ModbusReadRequestBlueprint> {
 
         private volatile @Nullable AtomicStampedValue<AsyncModbusReadResult> lastResult;
```

Comment:
Change in behaviour: errors are not cached -- we should cache them as well.

Reply:
Yes, I'm aware of this, and this is the main todo at the moment

Reply:
I proposed to take back changes from here https://github.com/openhab/openhab-addons/blob/2.5.x/bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/handler/ModbusPollerThingHandlerImpl.java#L75-L77

However, perhaps even simpler would be to 

1. Restore AtomicStampedKeyValue.java from https://github.com/openhab/openhab-addons/blob/2.5.x/bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/AtomicStampedKeyValue.java where key would be the request and value would be the response/error
2. Use single AtomicStampedKeyValue in poller, but erasing the types with Objects and casting where necessary when calling callbacks

Reply:
I have fixed this by adding a small private class to store the result+failure together. This way I could put back the original caching.

Reply:
Looks good! I added toString implementation to PollResult in https://github.com/mrbig/openhab2-addons/pull/1 since it is used in logging